### PR TITLE
How To Build A Custom URL Shortener Using Node.js, Express and MongoDB

### PR DESCRIPTION
**Pull request by marienjus**
### Article checklist
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by marienjus** [comment]
This fixes #1810

**Comment by marienjus** [comment]
cc: @cyberShaw

**Comment by cyberShaw** [comment]
Thank you for adding the author page, @marienjus!
I'll do the review shortly.

File: articles/node-url-shortener/index.md [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+This will download the named packages inside `node_modules` folder and update the `package.json` file with the dependencies.
+
+>>> Note: `nodemon` is a development dependency. To install this, run the command `npm install --save-dev nodemon`.
+
+
+### Setting Up Express server
+Inside our `URL-Shortener-Service` folder, create a file named `server.js`. The following is the initial code to start the Express server:
+
+```js
```
Comment (cyberShaw) [inline articles/node-url-shortener/index.md:41]: Please add code comments here. 
I know that you've explained the code, but comments make the code more readable.

File: articles/node-url-shortener/index.md [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+    app.listen(PORT, ()=> console.log('DB connection established'))
+```
+
+This starter code imports the express package. The `app = express()` creates an instance of our application.
+In our app, we need to listen to the incoming request. `app.listen` takes a port number and a callback function that is invoked upon a successful connection.
+
+### Configuring the MongoDB connection
+We will use the `mongoose` package that we installed via npm as the database driver. To configure the database, create a `config` folder inside our `URL-Shortener-Service` folder. Inside the `config` folder, add a file named `db.config.js`. This is the file to add the following database connection code:
+
```
Comment (cyberShaw) [inline articles/node-url-shortener/index.md:56]: Please add a heading that mentions that this code is to be included inside `db.config.js`

File: articles/node-url-shortener/index.md [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+```js
+const mongoose = require('mongoose')
+
+const URLSchema = new mongoose.Schema({
+    urlCode: String, 
+    longUrl: String,
+    shortUrl: String,
+    date:{type: String, default: Date.now}
+})
```
Comment (cyberShaw) [inline articles/node-url-shortener/index.md:84]: Please add code comments

File: articles/node-url-shortener/index.md [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+- `url.js`: This will be a POST route that takes an incoming request with the long URL and creates the short URL and inserts it into the database. Navigate in the `routes` folder and create a file named `url.js`.
+  
+- `redirect.js`: This is a GET for our URL redirects. It takes the short URL and redirects it to the actual long URL in the browser. To add this file, create a separate file named `redirect.js` inside the `routes` folder.
+
+#### The `url.js` POST Route 
+Now, let us create the POST route. Add the following code inside the `url.js` file:
+
+```js
+const express = require('express')
```
Comment (cyberShaw) [inline articles/node-url-shortener/index.md:110]: More code comments here would improve the quality of the article

File: articles/node-url-shortener/index.md [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+app.listen(PORT, console.log(`server started, listening PORT ${PORT}`))
+```
+
+To make our routes work, we have to use a [middleware pattern](https://expressjs.com/en/guide/using-middleware.html#middleware.application) is express. Middleware functions have access to request (req) and response object (res) in the application’s request-response cycle. This explains the `app.use()` methods. The first middleware allows our application to parse incoming request data format in JSON format. The `app.use('/', require('./routes/redirect'))` is the has the base URI that will configure the redirect route. In our POST route, the base URL is `/api/url` and the middleware as `app.use('/api/url', require('./routes/url'))`. Next, we need to test the application in postman.
+
+### A Working Demo In Postman
+
+![postman_test_gif](/engineering-education/node-url-shortener/postman.gif)
+
+### Conclusion
```
Comment (cyberShaw) [inline articles/node-url-shortener/index.md:229]: It would be helpful if you could upload the final code on GitHub and paste a link to the repository

**Review by cyberShaw — CHANGES_REQUESTED** [review]
Excellent article, @marienjus!

The article was really good and is good for a beginner to get started with node and express. I've included some comments, please make those changes and we can move on to the next round of review.

And please run your article through a grammar checker like Grammarly before submitting the article. there were a couple of minor changes which I fixed.

**Comment by marienjus** [comment]
Hey, @cyberShaw I hope everything is okay now. Done changes

**Comment by cyberShaw** [comment]
Great job on making the changes, @marienjus!
This article is ready for final review.

**Comment by hectorkambow** [comment]
plagiarism check done

**Comment by hectorkambow** [comment]
Great work @marienjus  
Seems like a very helpful aritcle.
🚀 
cc @cyberShaw

**Comment by marienjus** [comment]
Thank you. Hoping to write quality stuffs as I learn.

**Pull request merged**